### PR TITLE
Fix WM weekday holiday delay parsing

### LINKED ISSUE #7288: [Source Defect]: wm_com does not apply generic Labor Day weekday delay
### Source Name

`wm_com`

### What's happening?

The Waste Management source returns the normal Wednesday collection date instead of applying WM's one-day Labor Day delay.

For a residential US service address with a normal Wednesday pickup:

- WM `pickupinfo` currently returns `09-09-2026`.
- WM `/holidays` currently returns: `Residential: Labor Day is on Monday, September 7th, and we will be closed. Weekday collections will experience a delay of one day.`
- `Source.fetch()` nevertheless returns `2026-09-09`; the expected date is Thursday, `2026-09-10`.

This was reproduced with installed release `v2.32.0` and directly against current `master` commit `d0f2701fc7b7567991ea5dae55d8a18e84d74cc8`. The provider endpoints are responding successfully.

The current parser appears unable to apply this response for three reasons:

1. `_HOLIDAY_DATE_RE` only accepts slash-form numeric dates, not `September 7th`.
2. `_DELAY_RE` expects a digit before `day delay`, not `delay of one day`.
3. Even if the holiday date were parsed, `fetch()` only replaces a pickup whose exact date is a key in `holiday_map`. This WM message describes a week-wide delay for weekday collections, so a Wednesday pickup must move even though the holiday itself is Monday.

Potential fix: recognize WM's month-name/ordinal and word-number wording, and represent/apply generic holiday-week delays to affected weekday pickups after the holiday rather than only exact-date replacements.

### Error Logs

```Shell
No error is logged. The source fetch succeeds but returns the unadjusted date.
```

### Source Configuration

```YAML
source: wm_com
args:
  street: REDACTED
  city: REDACTED
  state: OH
  zip_code: REDACTED
  country: US
```

### Checklist

- [x] I have tested with the latest version (master branch); current master was fetched and `Source.fetch()` was run directly against the affected provider data.
- [x] The service provider's website/API is still working.
- [x] I have searched past issues (open and closed) for this problem.

### PR DESCRIPTION
**Pull request by Mike2541**
## Summary

- parse WM holiday dates in both numeric and named-month formats
- recognize WM's service-specific and generic weekday delay wording
- apply generic weekday delays from the holiday through Friday while preserving existing exact-date behavior
- add regression coverage for the existing numeric format, Labor Day, and Thanksgiving

Fixes #7288

## Testing

- `ruff check custom_components/waste_collection_schedule/waste_collection_schedule/source/wm_com.py tests/test_source_components.py`
- `python -m pytest tests -q` — 41 passed
- live `Source.fetch()` against the affected WM schedule returned `2026-09-10` instead of the unadjusted `2026-09-09`

### CODE REVIEW

File: custom_components/waste_collection_schedule/waste_collection_schedule/source/wm_com.py [hunk 1]
```diff
@@ -156,12 +167,24 @@ def _parse_holiday_message(
...
-            if raw.count("/") == 2:
-                dt = datetime.datetime.strptime(raw, "%m/%d/%Y")
+            if "/" in raw:
+                has_year = raw.count("/") == 2
+                if has_year:
+                    dt = datetime.datetime.strptime(raw, "%m/%d/%Y")
+                else:
+                    dt = datetime.datetime.strptime(raw, "%m/%d").replace(
+                        year=today.year
+                    )
```
[the remark this answers is not in the document]
Comment (Mike2541) [reply custom_components/waste_collection_schedule/waste_collection_schedule/source/wm_com.py:178]: Fixed in f7d82ca2. Numeric dates now distinguish two-digit years (`%y`) from four-digit years (`%Y`), and the regex accepts only those two supported year widths. Added a regression case for `11/24/26`.

File: custom_components/waste_collection_schedule/waste_collection_schedule/source/wm_com.py [hunk 2]
```diff
@@ -156,12 +167,24 @@ def _parse_holiday_message(
...
-                    dt = dt.replace(year=today.year + 1)
+                raw = re.sub(r"(\d)(?:st|nd|rd|th)\b", r"\1", raw)
+                has_year = "," in raw
+                date_format = "%B %d, %Y" if has_year else "%B %d"
+                dt = datetime.datetime.strptime(raw, date_format)
+                if not has_year:
+                    dt = dt.replace(year=today.year)
+
+            if not has_year and dt < today:
+                dt = dt.replace(year=today.year + 1)
```
[the remark this answers is not in the document]
Comment (Mike2541) [reply custom_components/waste_collection_schedule/waste_collection_schedule/source/wm_com.py]: Fixed in f7d82ca2. The rollover comparison now uses `dt.date()` against `date.today()`, so a no-year holiday occurring today remains in the current year. Added a frozen-today regression test.

**Comment by Mike2541** [comment]
Follow-up commit cf10f208 addresses an additional live edge case exposed on the day after a no-year holiday: WM keeps the holiday notice active during the affected collection week, so rolling the holiday into next year immediately after the holiday drops the delay. The parser now retains a just-passed no-year holiday for the collection week, with a regression test for Tuesday after Labor Day. Local pre-commit checks and the full suite pass (43 tests), and a live forced WM refresh restored the delayed Thursday collection date.

**Review by bbr111 — APPROVED** [review]
Thanks for this — and for the thorough follow-ups on the Copilot findings. I
re-checked both threads against the current tree and they're properly
resolved: the %y/%Y split matches what the regex now accepts, and the
rollover comparison is date-vs-date.

The cf10f208 week-window rule is the right call. WM keeps a notice live
through the affected collection week, so rolling the holiday forward a year
the morning after it passes silently drops the delay from exactly the
pickups that need it.

I pushed a few small tweaks rather than send this back round:

- Weekend-anchored holidays. `while collection_date.weekday() < 5` yields an
  empty mapping when the holiday itself lands on a Saturday or Sunday, so
  those notices produced no shifts at all. It now anchors to the observed
  weekday first — Saturday to the Friday before, Sunday to the Monday after,
  matching the US convention WM follows.

- Delay-length phrasing. `_DELAY_RE` only accepted a digit ahead of the word
  "delay", or the literal "one" after it, so "a delay of up to 2 days" was
  silently skipped. It now takes a digit or a small number word on either
  side. Both orderings are covered by tests, and a notice with no length at
  all still parses to no adjustment, same as before.

- `_parse_holiday_message` now takes an optional `today` argument. Your tests
  pinned the date with `patch.object(module.datetime, "date", ...)`, but since
  `wm_com` does a plain `import datetime`, `module.datetime` is the stdlib
  module — so that swaps out `datetime.date` for the whole process, not just
  the source under test. The rest of this test file patches attributes on the
  source module's own namespace, which is safe; this was the one outlier.
  Passing the date in keeps the tests deterministic without touching stdlib.
  The default is unchanged, so `_get_holiday_adjustments` behaves as before.

Also added a regression test for the stale-notice case: your tests covered
the "keep it in the current year" side of the new rule, but nothing asserted
that a genuinely year-less notice whose collection week has fully passed
still rolls forward. Easy to write now that the date is injectable.

42 passed on tests/test_source_components.py, ruff clean. Nice fix.
